olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+                          -1,  // denotes null-terminated string
+                          &s_out.front(), s_out.size());
+
+  if (conversion_result == 0) {
+    // Should not happen as 1st call have succeeded.
+    return false;
+  }
+  return true;
+}
+}  // namespace
+
+namespace olp {
+namespace network2 {
+
+#define LOGTAG "WinHttp"
+
+NetworkProtocolWinHttp::NetworkProtocolWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  recent_request_id_.store(1);
+  http_session_ = WinHttpOpen(L"NGMOS CLient", WINHTTP_ACCESS_TYPE_NO_PROXY,
+                              WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS,
+                              WINHTTP_FLAG_ASYNC);
+
+  if (!http_session_) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpOpen failed " << GetLastError());
+    return;
+  }
+
+  WinHttpSetStatusCallback(
+      http_session_,
+      (WINHTTP_STATUS_CALLBACK)&NetworkProtocolWinHttp::RequestCallback,
+      WINHTTP_CALLBACK_FLAG_ALL_COMPLETIONS | WINHTTP_CALLBACK_FLAG_HANDLES, 0);
+
+  event_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  // Store the memory tracking state during initialization so that it can be
+  // used by the thread.
+  thread_ = CreateThread(NULL, 0, NetworkProtocolWinHttp::Run, this, 0, NULL);
+  SetThreadPriority(thread_, THREAD_PRIORITY_ABOVE_NORMAL);
+}
+
+NetworkProtocolWinHttp::~NetworkProtocolWinHttp() {
+  std::vector<std::shared_ptr<ResultData>> pendingResults;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::vector<std::shared_ptr<RequestData>> requestsToCancel;
+    for (auto& request : http_requests_) {
+      requestsToCancel.push_back(request.second);
+    }
+    while (!requestsToCancel.empty()) {
+      std::shared_ptr<RequestData> request = requestsToCancel.front();
+      WinHttpCloseHandle(request->http_request);
+      request->http_request = NULL;
+      pendingResults.push_back(request->result_data);
+      requestsToCancel.erase(requestsToCancel.begin());
+    }
+  }
+
+  if (http_session_) {
+    WinHttpCloseHandle(http_session_);
+    http_session_ = NULL;
+  }
+
+  if (event_ != INVALID_HANDLE_VALUE) SetEvent(event_);
+  if (thread_ != INVALID_HANDLE_VALUE) {
+    if (GetCurrentThreadId() != GetThreadId(thread_)) {
+      WaitForSingleObject(thread_, INFINITE);
+    }
+  }
+  CloseHandle(event_);
+  CloseHandle(thread_);
+  thread_ = event_ = INVALID_HANDLE_VALUE;
+
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    http_connections_.clear();
+    while (!results_.empty()) {
+      pendingResults.push_back(results_.front());
+      results_.pop();
+    }
+  }
+
+  for (auto& result : pendingResults) {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    if (result->user_callback) {
+      result->user_callback(
+          NetworkResponse()
+              .WithRequestId(result->request_id)
+              .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
```

Comment:
```suggestion
              .WithStatus(static_cast<int>(ErrorCode::CANCELLED_ERROR))
```
I'm suggesting to use CANCELLED_ERROR, wdyt?